CMakeLists.txt

```diff
@@ -441,15 +441,106 @@ endif()
 
 # TODO: - ECC custom curves
 #       - Compressed key
-#       - CURVE25519
-#       - ED25519
-#       - CURVE448
-#       - ED448
 #       - FP ECC, fixed point cache ECC
 #       - ECC encrypt
 #       - PSK
 #       - Single PSK identity
 
+# CURVE25519
+set(WOLFSSL_CURVE25519_SMALL "no")
+set(WOLFSSL_CURVE25519_HELP_STRING "Enable Curve25519 (default: disabled)")
+set(WOLFSSL_CURVE25519 "no" CACHE STRING ${WOLFSSL_CURVE25519_HELP_STRING})
+set_property(CACHE WOLFSSL_CURVE25519 PROPERTY STRINGS "yes" "no" "small" "no128bit")
+
+if(WOLFSSL_OPENSSH)
+    override_cache(WOLFSSL_CURVE25519 "yes")
+endif()
+
+if(WOLFSSL_CURVE25519)
+    if("${WOLFSSL_CURVE25519}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DCURVE25519_SMALL")
+        set(WOLFSSL_CURVE25519_SMALL "yes")
+    endif()
+
+    if("${WOLFSSL_CURVE25519}" STREQUAL "no128bit" OR WOLFSSL_32BIT)
+        list(APPEND WOLFSSL_DEFINITIONS "-DNO_CURVED25519_128BIT")
+    endif()
+
+    list(APPEND WOLFSSL_DEFINITIONS "-DHAVE_CURVE25519")
+    set(WOLFSSL_FEMATH "yes")
+endif()
+
+# ED25519
+set(WOLFSSL_ED25519_SMALL "no")
+set(WOLFSSL_ED25519_HELP_STRING "Enable ED25519 (default: disabled)")
+set(WOLFSSL_ED25519 "no" CACHE STRING ${WOLFSSL_ED25519_HELP_STRING})
+set_property(CACHE WOLFSSL_ED25519 PROPERTY STRINGS "yes" "no" "small")
+
+if(WOLFSSL_OPENSSH)
+    override_cache(WOLFSSL_ED25519 "yes")
+endif()
+
+if(WOLFSSL_ED25519 AND NOT WOLFSSL_32BIT)
+    if("${WOLFSSL_ED25519}" STREQUAL "small" OR WOLFSSL_LOW_RESOURCE)
+        list(APPEND WOLFSSL_DEFINITIONS "-DED25519_SMALL")
+        set(WOLFSSL_ED25519_SMALL "yes")
+        set(WOLFSSL_CURVE25519_SMALL "yes")
+    endif()
+
+    if(NOT WOLFSSL_SHA512)
+        message(FATAL_ERROR "cannot enable ed25519 without enabling sha512.")
+    endif()
+
+    set(WOLFSSL_FEMATH "yes")
```

Comment:
Hmm, this does copy `configure.ac` logic, but only GEMATH is required to build and pass tests. Not sure who to ask about that. Probably best to include if we can't answer.

Reply:
Can you elaborate a bit on what you mean by "only GEMATH is required to build and pass tests"? WOLFSSL_FEMATH and WOLFSSL_GEMATH have the effect of controlling inclusion of certain source files in the generate_lib_src_list function of cmake/functions.cmake.

Reply:
When I first tried to build with ED25519, I missed these and was able to build with only `GEMATH`, so I don't think it's actually required. If you remove this line it still builds and tests pass.

Reply:
Right, but even if it builds and passes all our tests, the library itself is different without BUILD_FEMATH. If you run cmake with
```
cmake .. -DWOLFSSL_CURVE25519=yes -DWOLFSSL_ED25519=yes
```
and then run `make` and check the output, you won't see
```
Building C object CMakeFiles/wolfssl.dir/wolfcrypt/src/fe_low_mem.c.o
```
But with
```
cmake .. -DWOLFSSL_CURVE25519=small -DWOLFSSL_ED25519=small
```
you will, which is a consequence of BUILD_FEMATH being turned on.

Reply:
ok `small` must require it.